pkg/services/accesscontrol/middleware/middleware.go

```diff
@@ -1,68 +1,46 @@
 package middleware
 
 import (
-	"bytes"
 	"fmt"
 	"net/http"
-	"text/template"
 	"time"
 
-	"github.com/grafana/grafana/pkg/util"
-
-	macaron "gopkg.in/macaron.v1"
+	"gopkg.in/macaron.v1"
 
 	"github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/services/accesscontrol/eval"
+	"github.com/grafana/grafana/pkg/util"
 )
 
-func Middleware(ac accesscontrol.AccessControl) func(macaron.Handler, string, ...string) macaron.Handler {
-	return func(fallback macaron.Handler, permission string, scopes ...string) macaron.Handler {
+// TODO: add tests
+func Middleware(ac accesscontrol.AccessControl) func(macaron.Handler, eval.Evaluator) macaron.Handler {
+	return func(fallback macaron.Handler, evaluator eval.Evaluator) macaron.Handler {
 		if ac.IsDisabled() {
 			return fallback
 		}
 
 		return func(c *models.ReqContext) {
-			// We need this otherwise templated scopes get initialized only once, during the first call
-			runtimeScope := make([]string, len(scopes))
-			for i, scope := range scopes {
-				var buf bytes.Buffer
-
-				tmpl, err := template.New("scope").Parse(scope)
-				if err != nil {
-					c.JsonApiErr(http.StatusInternalServerError, "Internal server error", err)
-					return
-				}
-				err = tmpl.Execute(&buf, c.AllParams())
-				if err != nil {
-					c.JsonApiErr(http.StatusInternalServerError, "Internal server error", err)
-					return
-				}
-				runtimeScope[i] = buf.String()
-			}
-
-			hasAccess, err := ac.Evaluate(c.Req.Context(), c.SignedInUser, permission, runtimeScope...)
-			if err != nil {
-				Deny(c, permission, runtimeScope, err)
+			if err := evaluator.Inject(c.AllParams()); err != nil {
+				c.JsonApiErr(http.StatusInternalServerError, "Internal server error", err)
 				return
 			}
-			if !hasAccess {
-				Deny(c, permission, runtimeScope, nil)
+
+			hasAccess, err := ac.Evaluate(c.Req.Context(), c.SignedInUser, evaluator)
+			if !hasAccess || err != nil {
+				Deny(c, err)
 				return
 			}
 		}
 	}
 }
 
-func Deny(c *models.ReqContext, permission string, scopes []string, err error) {
+func Deny(c *models.ReqContext, err error) {
 	id := newID()
 	if err != nil {
 		c.Logger.Error("Error from access control system", "error", err, "accessErrorID", id)
 	} else {
-		c.Logger.Info("Access denied",
-			"userID", c.UserId,
-			"permission", permission,
-			"scopes", scopes,
-			"accessErrorID", id)
+		c.Logger.Info("Access denied", "userID", c.UserId, "accessErrorID", id)
```

Comment:
I think this was verbose on purpose, for the administrators to see what permission on which scope the user might be lacking. Could we have another way to access this piece of information?

Reply:
If we want to keep that we need to Add a way to get underlying permissions from the evaluator interface. That would not be hard to add